### PR TITLE
WIP: Add the kvm serverspec

### PR DESCRIPTION
**Pull request by vlope093**
Dependent on PR #1554

### CODE REVIEW

File: roles/common/templates/etc/serverspec/kvm_spec.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  file_contents = ['#listen_tcp = 1',
+                   'listen_tcp = 1',
+                   '#listen_tls = 0',
+                   'listen_tls = 0']
+  file_contents.each do |file_line|
+    it { should contain file_line }
+  end
+end
+
+describe command('cat /etc/libvirt/libvirtd.conf|grep ^listen_') do
```
Comment (pwplusnick) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: Use the file resource type.

File: roles/common/templates/etc/serverspec/kvm_spec.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  file_contents =["listen_tls = 0", "\nlisten_tcp = 1\n"].join()
+  its(:stdout) { should eq file_contents }
+end
```
Comment (j2sol) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: Isn't this check the same as the check above it? Why are you checking for listen_tls and listen_tcp twice?

File: roles/common/templates/etc/serverspec/kvm_spec.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+describe file('/etc/libvirt/qemu.conf') do
+  file_contents = ['#vnc_listen = "0.0.0.0"',
+                   '#       user = "qemu"   # A user named "qemu"',
+                   '#       user = "+0"     # Super user (uid=0)',
+                   '#       user = "100"    # A user named "100" or a user with uid=100',
+                   '#user = "root"']
+  file_contents.each do |file_line|
```
Comment (j2sol) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: So what's the point of this check? To make sure vnc isn't listening? What if there was an uncommented version of these lines as well?

File: roles/rabbitmq/tasks/main.yml [hunk 4]
```diff
@@ -46,3 +46,9 @@
 
 - include: monitoring.yml tags=monitoring,common
   when: monitoring.enabled|default('True')|bool
+ 
+- include: serverspec.yml
+  tags: 
+    - serverspec
+  when: serverspec.enabled|default('False')|bool
```
Comment (j2sol) [inline roles/rabbitmq/tasks/main.yml]: This change seems unrelated to KVM serverspec.

File: roles/rabbitmq/tasks/serverspec.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+- name: serverspec checks for kvm
+  template: src={{ item }}
+            dest=/etc/serverspec/spec/localhost/
+            mode=0755
+  with_fileglob: ../templates/etc/serverspec/*
```
Comment (j2sol) [inline roles/rabbitmq/tasks/serverspec.yml]: This change seems unrelated to kvm serverspec.

File: roles/common/templates/etc/serverspec/kvm_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
+require 'spec_helper'
```
Comment (j2sol) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: Why is a KVM server spec going into the common role? There are systems that get the common role which are not compute nodes, thus wouldn't have KVM bits running. Perhaps this was meant for the nova-data role?

Comment (pwplusnick) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: Agreed. nova-data seems more fitting.

File: roles/common/templates/etc/serverspec/kvm_spec.rb (hunk 1 above)
Comment (vlope093) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: We are not using the file contain resource type since it will be looking for 'listen_tcp =1" in the file and it may take the comment (#listen_tcp =1) as a valid string.

File: roles/common/templates/etc/serverspec/kvm_spec.rb (hunk 2 above)
Comment (vlope093) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: yes, I'll fix that.

File: roles/common/templates/etc/serverspec/kvm_spec.rb (hunk 1 above)
Comment (j2sol) [inline roles/common/templates/etc/serverspec/kvm_spec.rb]: The file resource appears to support match regex, in which you can say that the line must start with listen_tcp rather than just have listen_tcp somewhere in it, at least according to my reading of http://serverspec.org/resource_types.html#file

File: roles/common/tasks/main.yml [hunk 7]
```diff
@@ -114,9 +114,3 @@
     - monitoring
     - common
   when: monitoring.enabled|default('True')|bool
-
-- include: audit-logging.yml
-  tags:
-    - logrotate
-    - logging
-  when: logging.enabled|default('True')|bool
```
Comment (j2sol) [inline roles/common/tasks/main.yml]: Why is this change here? Seems unrelated.

File: roles/nova-data/templates/etc/severspec/kvm_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
+require 'spec_helper'
+
+describe file('/var/lib/libvirt/images') do
```
Comment (j2sol) [inline roles/nova-data/templates/etc/severspec/kvm_spec.rb]: I'm not sure we actually use this directory for anything. I mean, it doesn't hurt to look at it, but I don't expect it to ever be used.

File: roles/rabbitmq/tasks/main.yml [hunk 9]
```diff
@@ -45,4 +45,4 @@
   file: group=root owner=root mode=0755 path=/usr/local/bin/rabbitmqadmin
 
 - include: monitoring.yml tags=monitoring,common
-  when: monitoring.enabled|default('True')|bool
+  when: monitoring.enabled|default('True')|bool
```
Comment (j2sol) [inline roles/rabbitmq/tasks/main.yml]: Added whitespace here.

File: roles/nova-data/templates/etc/serverspec/kvm_spec.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+describe file('/dev/kvm') do
+  it { should exist }
+  it { should be_mode 660 }
+end 
+
+describe file('/etc/libvirt/libvirtd.conf') do
+  file_contents = ['unix_sock_group = "libvirtd"',
+                   'unix_sock_ro_perms = "0777"',
+                   'unix_sock_rw_perms = "0770"']
```
Comment (j2sol) [inline roles/nova-data/templates/etc/serverspec/kvm_spec.rb]: I would say these need to check if the line _starts with_ these items, otherwise it'll catch commented out lines.

File: roles/nova-data/templates/etc/serverspec/kvm_spec.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  file_contents.each do |file_line|
+    it { should contain file_line }
+  end
+end
+
+describe file('/etc/libvirt/libvirtd.conf') do
+  it { should contain "^listen_tls = 0" }
+  it { should contain "^listen_tcp = 1" }
```
Comment (j2sol) [inline roles/nova-data/templates/etc/serverspec/kvm_spec.rb]: Why aren't these two bundled with the above contents?

File: roles/nova-data/templates/etc/serverspec/kvm_spec.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+end
+
+describe file('/etc/libvirt/qemu.conf') do
+  file_contents = ['#vnc_listen = "0.0.0.0"',
+                   '#       user = "qemu"   # A user named "qemu"',
+                   '#       user = "+0"     # Super user (uid=0)',
+                   '#       user = "100"    # A user named "100" or a user with uid=100',
+                   '#user = "root"']
+  file_contents.each do |file_line|
+    it { should contain file_line }
```
Comment (j2sol) [inline roles/nova-data/templates/etc/serverspec/kvm_spec.rb]: Seems like you're checking to make sure that vnc_listen has been commented out. But instead shoudn't you check to make sure there aren't any lines that start with `vnc_listen` that are not commented out?

File: roles/nova-data/templates/etc/serverspec/kvm_spec.rb (hunk 11 above)
Comment (pwplusnick) [inline roles/nova-data/templates/etc/serverspec/kvm_spec.rb]: That kind of thing usually do to it being a different rule in the techspec (read: ITCS104). When we label them in the future with the yaml techspec definitions, they will be two separate checks.

File: roles/nova-data/templates/etc/serverspec/kvm_spec.rb (hunk 10 above)
Comment (pwplusnick) [inline roles/nova-data/templates/etc/serverspec/kvm_spec.rb]: Agreed.

File: roles/nova-data/templates/etc/serverspec/kvm_spec.rb (hunk 12 above)
Comment (pwplusnick) [inline roles/nova-data/templates/etc/serverspec/kvm_spec.rb]: Probably an artifact of screen scraping if I had to guess. Should be removed.

**Comment by j2sol** [comment]
ok to test

**Comment by j2sol** [comment]
retest
